tools/update.php

```diff
@@ -0,0 +1,747 @@
+    define('MAJOR', 0);
+    define('MINOR', 1);
+    define('BUGFIX', 2);
+    // Convert string representation of version into arrays of ints
+    // TODO: check for dev branches
+    $from_versions = array_map('intval', explode('.', $version_from));
+    $to_versions   = array_map('intval', explode('.', $version_to));
+    // Calculate difference between old version and latest
+    $diff_major = $to_versions[MAJOR] - $from_versions[MAJOR];
+    $diff_minor = $to_versions[MINOR] - $from_versions[MINOR];
+    if ($diff_major < 0) {
+        echo '[!] Your version of LORIS is ahead of the latest release. If you'
+        . ' are not a developer working on the bleeding edge of LORIS, then '
+        . 'something has gone very wrong. No SQL patches will be displayed.'
+        . PHP_EOL;
+        return array();
+    }
+    echo "[**] Latest version $version_to is ahead of installed $version_from "
+        . "by $diff_major MAJOR release(s), $diff_minor MINOR release(s)."
+        . PHP_EOL;
+
+    // For every major version released between the version that is installed
+    // and the latest version, add the relevant patches if they begin with
+    // a number in that range. For example, if upgrading from 17 to 19, the
+    // array will contain all aptches beginning with "17", "18", or "19".
+    // i.e.:
+    //   [0] => /var/www/loris/SQL/Release_patches/17.0_To_18.0_upgrade_A.sql
+    //   [1] => /var/www/loris/SQL/Release_patches/17.0_To_18.0_upgrade_B.sql
+    //   [2] => /var/www/loris/SQL/Release_patches/18.0_To_19.0_upgrade.sql
+    //   [3] => /var/www/loris/SQL/Release_patches/19.0_To_19.1_upgrade.sql
+    // Range will run once when $start = $end. We want this in the case where
+    // there is no difference in major versions but there is a change in minor
+    // versions. The below loops will, in this case, match the minor release
+    // patches.
+    $start = $from_versions[MAJOR];
+    $end   = $start; // default case: there is a minor release difference
+    if ($diff_major > 0) {
+        $end = $to_versions[MAJOR];
+    }
+    $patch_dir           = $loris_root . 'SQL/Release_patches/';
+    $all_release_patches = glob($patch_dir . '*.sql');
+    $patches = [];
+    if ($diff_major > 0 || $diff_minor > 0) {
+        foreach (range($start, $end) as $v) {
+            foreach ($all_release_patches as $patch) {
+                // strpos must be 0 or else this will match w/ e.g. 16.0_To_17.0
+                // when the server is already on version 17
+                if (strpos(basename($patch), "$v") === 0) {
+                    $patches[] = $patch;
+                }
+            }
+        }
+    }
+    if (isDev()) {
+        echo "[+] Developer instance detected. Including developer patches..."
+            . PHP_EOL;
+
+        // Add all patches in Archive/$MAJOR.$MINOR. Everything other needed
+        // command will be in the Release patches which are been added above
```

Comment:
the release patch is the concatenation of the patches in `Archive` why are you readding them here ?

Reply:
Right... I guess this should be `New_patches`. I remember at some point we were actually putting new patches in `Archive` so I based this on that.

I'll update this.

Reply:
lets talk